pi5/pinctrl_utils.go

```diff
@@ -0,0 +1,231 @@
+returns:
+
+	numPAddrCells:	number of 32 bit chunks needed to represent the parent address
+	numAddrSpaceCells:	number of 32 bit chunks needed to represent the size of the parent address space
+*/
+func getNumAddrSizeCellsInfo(parentNodePath string) (uint32, uint32, error) {
+
+	// get #address - cells info for child node using the parent Node
+	npaByteContents, err := os.ReadFile(parentNodePath + "/#address-cells")
+	if err != nil {
+		return 0, 0, fmt.Errorf("trouble getting addr cells info for %s: %w\n", parentNodePath, err)
+	}
+	numPAddrCells := binary.BigEndian.Uint32(npaByteContents[:4])
+
+	// get #size - cells info for child node using the parent Node
+	npsByteContents, err := os.ReadFile(parentNodePath + "/#size-cells")
+	if err != nil {
+		return 0, 0, fmt.Errorf("trouble getting size cells info for %s: %w\n", parentNodePath, err)
+	}
+	numAddrSpaceCells := binary.BigEndian.Uint32(npsByteContents[:4]) // reading 4 bytes because the number is represented by 1 uint32. 4bytes * 8bits/byte = 32 bits
+	return numPAddrCells, numAddrSpaceCells, err
+}
+
+// Reads the /reg file and converts the bytestream into a uint64 representing the GPIO Node's physical address within its parent's space. (pre address mapping)
+func getRegAddr(childNodePath string, numPAddrCells uint32) (uint64, error) {
+
+	//newPath := "/proc/device-tree/axi/pcie@120000/rp1/gpio@d0000/reg"
+
+	childNodePath += "/reg"
+	childNodePath = strings.TrimSpace(childNodePath)
+
+	re := regexp.MustCompile(`[\x00-\x1F\x7F-\x9F]`) // gets rid of random non printable chars. works for now but make cleaner later
+	childNodePath = re.ReplaceAllString(childNodePath, "")
+
+	regByteContents, err := os.ReadFile(childNodePath)
+	if err != nil {
+		return INVALID_ADDR, fmt.Errorf("trouble getting reg addr info for %s: %w\n", childNodePath, err)
+	}
+
+	physAddr := INVALID_ADDR
+	if numPAddrCells == 1 { // reading in 32 bits. regardless we must convert to a 64 bit address so we add a bunch of 0s to the beginning.
+		physAddr = uint64(binary.BigEndian.Uint32(regByteContents[:(4 * numPAddrCells)]))
+	} else if numPAddrCells == 2 { // reading in 64 bits
+		physAddr = binary.BigEndian.Uint64(regByteContents[:(4 * numPAddrCells)])
+	} else { // reading in more than 64 bits. we only want the last 64 bits of the address though, so we cut off the other portion of it
+		physAddr = binary.BigEndian.Uint64(regByteContents[(4 * (numPAddrCells - 2)):(4 * numPAddrCells)])
+	}
+	return physAddr, err
+}
+
+// Reads the /ranges file and converts the bytestream into integers representing the < child address parent address parent size >
+func getRangesAddrInfo(childNodePath string, numCAddrCells uint32, numPAddrCells uint32, numAddrSpaceCells uint32) ([]rangeInfo, error) {
+
+	childNodePath += "/ranges"
+	childNodePath = strings.TrimSpace(childNodePath)
+	re := regexp.MustCompile(`[\x00-\x1F\x7F-\x9F]`) // gets rid of random non printable chars. works for now but make cleaner later. remvonig the -9F causes bugs.
```

Comment:
make a helper function to remove the nonprintable chars so you dont have to copy and paste these lines